dag_in_context/src/pretty_print.rs

```diff
@@ -0,0 +1,831 @@
+        let mut termdag = TermDag::default();
+        let (sort, value) = egraph
+            .eval_expr(&egglog::ast::Expr::Var((), "EXPR___".into()))
+            .unwrap();
+        let (_, extracted) = egraph.extract(value, &mut termdag, &sort);
+        let mut converter = FromEgglog {
+            termdag: &termdag,
+            conversion_cache: HashMap::new(),
+        };
+        Ok(Self::from_expr(converter.expr_from_egglog(extracted)))
+    }
+
+    pub fn from_expr(expr: RcExpr) -> PrettyPrinter {
+        PrettyPrinter {
+            expr,
+            symbols: indexmap::IndexMap::new(),
+            log: vec![],
+            table: std::collections::BTreeMap::new(),
+        }
+    }
+
+    // accept new expr and preserve symbols, but clear the log
+    pub fn new_expr(&mut self, expr: RcExpr) {
+        self.expr = expr;
+        self.log = vec![];
+    }
+
+    pub fn to_egglog_default(&mut self, binding: String) -> String {
+        self.to_egglog(&|rc, len| (rc > 1 && len > 30) || len > 80, binding)
+    }
+
+    // turn the Expr to a nested egglog with intermediate variables.
+    // fold_when: provide a function that decide when to fold the egglog expression to a let binding
+    pub fn to_egglog(
+        &mut self,
+        fold_when: &dyn Fn(usize, usize) -> bool,
+        binding: String,
+    ) -> String {
+        self.assign_fresh_var(&self.expr.clone());
+        let res = self.fold_expr(&self.expr.clone(), fold_when, false);
+        let log = self
+            .log
+            .iter()
+            .map(|fresh_var| {
+                let symbol = self.table.get(fresh_var).unwrap();
+                let pretty = symbol.symbol_log_to_str(false);
+                format!("(let {fresh_var} \n{pretty})\n")
+            })
+            .collect::<Vec<_>>()
+            .join("");
+        log + &format!("\n(let {binding} \n{})\n", res.pretty())
+    }
+
+    pub fn to_rust_default(&mut self, binding: String) -> String {
+        self.to_rust(
+            &|rc, len| (rc > 1 && len > 30) || rc > 4 || len > 80,
+            binding,
+        )
+    }
+
+    // turn the Expr to a rust ast macro string.
+    // fold_when: provide a function that decide when to fold the macro to a let binding
+    pub fn to_rust(&mut self, fold_when: &dyn Fn(usize, usize) -> bool, binding: String) -> String {
+        self.assign_fresh_var(&self.expr.clone());
+        let res = self.fold_expr(&self.expr.clone(), fold_when, false);
+        let log = self
+            .log
+            .iter()
+            .map(|fresh_var| {
+                let symbol = self.table.get(fresh_var).unwrap();
+                let ast = symbol.symbol_log_to_str(true);
+                format!("let {fresh_var} = {ast};")
+            })
+            .collect::<Vec<_>>()
+            .join("\n");
+        log + &format!("\nlet {binding} = {};\n", res.to_ast())
+    }
+
+    fn assign_fresh_var(&mut self, expr: &RcExpr) {
```

Comment:
What does this method do? Needs a doc comment

Reply:
I think I understand what this method does now- it gives a fresh variable to each node in the AST.
However, you are already traversing the AST in the `fold_expr` method.
You can simply give new variables there instead of here, so you don't need this method